### PR TITLE
Move to have estimate be a list column of data.frames for dynamic survival metrics

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
This PR makes it so the input to dynamic survival metrics, looks similar to the output that comes out when we predict censored models.

``` r
# pak::pak("tidymodels/yardstick@list-column-survival")
library(yardstick)

lung_surv
#> # A tibble: 228 × 3
#>    .pred            .pred_time surv_obj
#>    <list>                <dbl>   <Surv>
#>  1 <tibble [5 × 5]>       324.      306
#>  2 <tibble [5 × 5]>       476.      455
#>  3 <tibble [5 × 5]>       521.     1010
#>  4 <tibble [5 × 5]>       368.      210
#>  5 <tibble [5 × 5]>       506.      883
#>  6 <tibble [5 × 5]>       324.     1022
#>  7 <tibble [5 × 5]>       361.      310
#>  8 <tibble [5 × 5]>       353.      361
#>  9 <tibble [5 × 5]>       379.      218
#> 10 <tibble [5 × 5]>       254.      166
#>    .pred            .pred_time        1
#>    <list>                <dbl>        1
#>  1 <tibble [5 × 5]>       324.        0
#>  2 <tibble [5 × 5]>       476.        1
#>  3 <tibble [5 × 5]>       521.        1
#>  4 <tibble [5 × 5]>       368.        0
#>  5 <tibble [5 × 5]>       506.        1
#>  6 <tibble [5 × 5]>       324.        1
#>  7 <tibble [5 × 5]>       361.        1
#>  8 <tibble [5 × 5]>       353.        1
#> # ℹ 218 more rows

lung_surv$.pred[[1]]
#> # A tibble: 5 × 5
#>   .eval_time .pred_survival .weight_time .pred_censored .weight_censored
#>        <dbl>          <dbl>        <dbl>          <dbl>            <dbl>
#> 1        100          0.819         100.          0.995             1.01
#> 2        200          0.597         200.          0.928             1.08
#> 3        300          0.407         300.          0.760             1.31
#> 4        400          0.264         306.          0.735             1.36
#> 5        500          0.164         306.          0.735             1.36

brier_survival(
  data = lung_surv,
  truth = surv_obj,
  estimate = .pred
)
#> # A tibble: 5 × 4
#>   .metric        .estimator .eval_time .estimate
#>   <chr>          <chr>           <dbl>     <dbl>
#> 1 brier_survival standard          100     0.109
#> 2 brier_survival standard          200     0.194
#> 3 brier_survival standard          300     0.219
#> 4 brier_survival standard          400     0.222
#> 5 brier_survival standard          500     0.197

mixed_set <- metric_set(brier_survival, concordance_survival)

mixed_set(
  data = lung_surv,
  truth = surv_obj,
  estimate = .pred,
  pred_time = .pred_time
)
#> # A tibble: 6 × 4
#>   .metric              .estimator .eval_time .estimate
#>   <chr>                <chr>           <dbl>     <dbl>
#> 1 brier_survival       standard          100     0.109
#> 2 brier_survival       standard          200     0.194
#> 3 brier_survival       standard          300     0.219
#> 4 brier_survival       standard          400     0.222
#> 5 brier_survival       standard          500     0.197
#> 6 concordance_survival standard           NA     0.637
```

<sup>Created on 2023-03-31 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/metrics.R [hunk 1]
```diff
@@ -179,8 +178,8 @@ metrics.data.frame <- function(data,
 #' predictions (the class probability columns) as bare column names or
 #' `tidyselect` selectors to `...`.
 #'
-#' When mixing dynamic and static survival metrics, `censoring_weights` and
-#' `eval_time` is only needed when dynamic metrics are included in the set.
+#' When mixing dynamic and static survival metrics, `pred_time` is used to
```
Comment (topepo) [inline R/metrics.R:181]: I'd like to change this to be similar to how it works with classification: 

```r
lung_surv |>
  concordance_survival(truth = surv_obj, estimate = .pred_time)

lung_surv |>
  brier_survival(truth = surv_obj,  .pred_survival)

lung_surv |>
  mixed_set(truth = surv_obj, estimate = .pred_time,  .pred_survival)
```

IMO this is most analogous to the classification scenario where `estimate` is the point estimate (hard class prediction / time prediction) and ... maps to the probability estimates (class probabilities / survival prediction).  Not a perfect analogy due the static and dynamic split but I think that it works better.

File: R/surv-roc_survival_auc.R [hunk 2]
```diff
@@ -7,7 +7,7 @@
 #' @template return-dynamic-survival
 #' @details
 #'
-#' This method automatically groups by the `eval_time` argument.
+#' This method automatically groups by the `.eval_time` values.
```
Comment (topepo) [inline R/surv-roc_survival_auc.R:10]: This is a dynamic metric. 

``` r
library(yardstick)

roc_auc_survival(
  lung_surv,
  truth = surv_obj,
  estimate = .pred
)
#> # A tibble: 1 × 3
#>   .metric          .estimator .estimate
#>   <chr>            <chr>          <dbl>
#> 1 roc_auc_survival standard       0.997
```

<sup>Created on 2023-04-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by topepo** [comment]
For metrics requiring the IPCW, can we check earlier and give a more informative message? 

``` r
library(tidymodels)
```

``` r

tidymodels_prefer()
options(pillar.advice = FALSE, pillar.min_title_chars = Inf)
```

``` r

lung_surv %>% 
  mutate(
    .pred = map(.pred, ~ dplyr::select(.x, -.weight_censored))
  ) %>% 
  brier_survival(
    truth = surv_obj,
    estimate = .pred
  )
#> Error in `dplyr::reframe()` at yardstick/R/template.R:369:2:
#> ℹ In argument: `.estimate = fn(...)`.
#> ℹ In group 1: `.eval_time = 100`.
#> Caused by error in `dplyr::summarize()`:
#> ℹ In argument: `.estimate = brier_survival_impl(...)`.
#> ℹ In group 1: `.eval_time = 100`.
#> Caused by error in `brier_survival_impl()`:
#> ! object '.weight_censored' not found
#> Backtrace:
#>      ▆
#>   1. ├─... %>% brier_survival(truth = surv_obj, estimate = .pred)
#>   2. ├─yardstick::brier_survival(., truth = surv_obj, estimate = .pred)
#>   3. ├─yardstick:::brier_survival.data.frame(., truth = surv_obj, estimate = .pred) at yardstick/R/surv-brier.R:52:2
#>   4. │ └─yardstick::dynamic_survival_metric_summarizer(...) at yardstick/R/surv-brier.R:68:2
#>   5. │   ├─dplyr::reframe(...) at yardstick/R/template.R:369:2
#>   6. │   └─dplyr:::reframe.data.frame(...)
#>   7. │     └─dplyr:::summarise_cols(.data, dplyr_quosures(...), by, "reframe")
#>   8. │       ├─base::withCallingHandlers(...)
#>   9. │       └─dplyr:::map(quosures, summarise_eval_one, mask = mask)
#>  10. │         └─base::lapply(.x, .f, ...)
#>  11. │           └─dplyr (local) FUN(X[[i]], ...)
#>  12. │             └─mask$eval_all_summarise(quo)
#>  13. │               └─dplyr (local) eval()
#>  14. ├─yardstick (local) fn(...)
#>  15. │ └─... %>% ... at yardstick/R/surv-brier.R:106:2
#>  16. ├─dplyr::summarize(...)
#>  17. ├─dplyr:::summarise.grouped_df(...)
#>  18. │ └─dplyr:::summarise_cols(.data, dplyr_quosures(...), by, "summarise")
#>  19. │   ├─base::withCallingHandlers(...)
#>  20. │   └─dplyr:::map(quosures, summarise_eval_one, mask = mask)
#>  21. │     └─base::lapply(.x, .f, ...)
#>  22. │       └─dplyr (local) FUN(X[[i]], ...)
#>  23. │         └─mask$eval_all_summarise(quo)
#>  24. │           └─dplyr (local) eval()
#>  25. ├─yardstick:::brier_survival_impl(...)
#>  26. └─base::.handleSimpleError(...) at yardstick/R/surv-brier.R:135:2
#>  27.   └─dplyr (local) h(simpleError(msg, call))
#>  28.     └─dplyr (local) handler(cnd)
#>  29.       └─rlang::abort(message, class = error_class, parent = parent, call = error_call)
```

<sup>Created on 2023-04-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Review by topepo — CHANGES_REQUESTED** [review]

**Comment by topepo** [comment]
One other thing (for this or a separate PR). The integrated score metric(s) should probably have a different class; say `"integrated_survival_metric"`.

**Comment by EmilHvitfeldt** [comment]
@topepo For the validation of surv input, is it expected that the input will include these columns:
```r
  .eval_time .pred_survival .weight_time .pred_censored .weight_censored
```

File: R/metrics.R (hunk 1 above)
Comment (EmilHvitfeldt) [reply R/metrics.R]: Done!

``` r
library(yardstick)

lung_surv |>
  concordance_survival(truth = surv_obj, estimate = .pred_time)
#> # A tibble: 1 × 3
#>   .metric              .estimator .estimate
#>   <chr>                <chr>          <dbl>
#> 1 concordance_survival standard       0.637

lung_surv |>
  brier_survival(truth = surv_obj,  .pred)
#> # A tibble: 5 × 4
#>   .metric        .estimator .eval_time .estimate
#>   <chr>          <chr>           <dbl>     <dbl>
#> 1 brier_survival standard          100     0.109
#> 2 brier_survival standard          200     0.194
#> 3 brier_survival standard          300     0.219
#> 4 brier_survival standard          400     0.222
#> 5 brier_survival standard          500     0.197

mixed_set <- metric_set(concordance_survival, brier_survival)

lung_surv |>
  mixed_set(truth = surv_obj, estimate = .pred_time,  .pred)
#> # A tibble: 6 × 4
#>   .metric              .estimator .eval_time .estimate
#>   <chr>                <chr>           <dbl>     <dbl>
#> 1 brier_survival       standard          100     0.109
#> 2 brier_survival       standard          200     0.194
#> 3 brier_survival       standard          300     0.219
#> 4 brier_survival       standard          400     0.222
#> 5 brier_survival       standard          500     0.197
#> 6 concordance_survival standard           NA     0.637
```

<sup>Created on 2023-04-03 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by topepo** [comment]
`.weight_time` and  `.pred_censored` are not required

**Comment by EmilHvitfeldt** [comment]
> For metrics requiring the IPCW, can we check earlier and give a more informative message?

Checking is now done earlier with more informative error message. It is still not ideal since we haven't gotten around to passing the call yet, re: https://github.com/tidymodels/yardstick/issues/348. I will do that next

``` r
library(tidymodels)

lung_surv %>% 
  mutate(
    .pred = map(.pred, ~ dplyr::select(.x, -.weight_censored))
  ) %>% 
  brier_survival(
    truth = surv_obj,
    .pred
  )
#> Error in `dplyr::reframe()`:
#> ℹ In argument: `.estimate = fn(...)`.
#> Caused by error in `validate_surv_truth_list_estimate()`:
#> ! All data.frames of `estimate` should include column names: `.eval_time`, `.pred_survival`, and `.weight_censored`.
```

<sup>Created on 2023-04-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Pull request merged**